clusterctl/providercomponents/providercomponents.go

```diff
@@ -0,0 +1,98 @@
+package providercomponents
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	core "k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+const (
+	configMapName                  = "clusterctl"
+	configMapProviderComponentsKey = "provider-components"
+)
+
+type Store struct {
+	// If present the provider components will be loaded from and saved to this file
+	ExplicitPath string
+	// If present and ExplicitPath is not present, provider components will be loaded and saved to this store
+	ConfigMap v1.ConfigMapInterface
+}
+
+func NewFromConfigMap(configMap v1.ConfigMapInterface) (*Store, error) {
+	store := Store{
+		ConfigMap: configMap,
+	}
+	return &store, nil
+}
+
+func NewFromClientset(clientset *kubernetes.Clientset) (*Store, error) {
+	return NewFromConfigMap(clientset.CoreV1().ConfigMaps(core.NamespaceDefault))
+}
+
+func (pc *Store) Save(providerComponents string) error {
+	if pc.ExplicitPath == "" {
+		return pc.saveToConfigMap(providerComponents)
+	}
+	return ioutil.WriteFile(pc.ExplicitPath, []byte(providerComponents), 0644)
+}
+
+func (pc *Store) Load() (string, error) {
+	if pc.ExplicitPath == "" {
+		return pc.loadFromConfigMap()
+	}
+	return pc.loadFromFile()
+}
+
+func (pc *Store) loadFromFile() (string, error) {
+	bytes, err := ioutil.ReadFile(pc.ExplicitPath)
+	if err != nil {
+		return "", fmt.Errorf("error when loading provider components from '%v': %v", pc.ExplicitPath, err)
+	}
+	return string(bytes), nil
+}
+
+func (pc *Store) saveToConfigMap(providerComponents string) error {
+	configMap, err := pc.ConfigMap.Get(configMapName, meta.GetOptions{})
+	if err != nil {
```

Comment:
should we ignore all errors or just the not found error?
might be useful to log non-not found errors.

you can use k8s.io/apimachinery/pkg/api/errors and errors.IsNotFound(err) to check

Reply:
Thank you for telling me about this! IsNotFound is definitely the only error that I wanted to handle.